## Designing and Creating a Database

In this project, we will cover:

- Import data into SQLite
- Design a normalized database schema
- Create tables for our schema
- Insert data into our schema

We will be working with a file of [Major League Baseball] games from [Retrosheet]. Retrosheet compiles detailed statistics on baseball games from the 1800s through to today. The main file we will be working from `game_log.csv`, has been produced by combining 127 separate CSV files from retrosheet, and has been pre-cleaned to remove some inconsistencies. The game log has hundreds of data points on each game which we will normalize into several separate tables using SQL, providing a robust database of game-level statistics.

In addition to the main file, we have also included three 'helper' files, also sourced from Retrosheet:

- `park_codes.csv`
- `person_codes.csv`
- `team_codes.csv`

These three helper files in some cases contain extra data, but will also make things easier as they will form the basis for three of our normalized tables.

An important first step when working with any new data is to perform exploratory data analysis (EDA). EDA gets us familiar with the data and gives us a level of background knowledge that will help us throughout our project. The methods you use when performing EDA will depend on what you plan to do with the data. In our case, we're wanting to create a normalized database, so our focus should be:

- Becoming familiar, at a high level, with the meaning of each column in each file.
- Thinking about the relationships between columns within each file.
- Thinking about the relationships between columns across different files.

A `game_log_fields.txt` file is included from Retrosheet which explains the fields included in our main file, which will be useful to assist our EDA.


[//]: # (Reference Links)

[Major League Baseball]: https://en.wikipedia.org/wiki/Major_League_Baseball
[Retrosheet]: http://www.retrosheet.org/

## Getting to Know the Data

In [1]:
# import libraries
import pandas as pd
import sqlite3
import csv

In [2]:
# set pandas options to ensure the DataFrame output isn't truncated

pd.set_option('max_columns', 180)
pd.set_option('max_rows', 200000)
pd.set_option('max_colwidth', 5000)

In [3]:
# read files

game_log = pd.read_csv('game_log.csv', low_memory=False)
park = pd.read_csv('park_codes.csv')
person = pd.read_csv('person_codes.csv')
team = pd.read_csv('team_codes.csv')

In [4]:
dfs = {
    'game_log': game_log,
    'park': park,
    'person': person,
    'team': team,
}

In [5]:
for name in dfs.keys():
    print(name)
    df = dfs[name]
    print(f"Size is {df.shape}")
    print('Columns:', df.columns)
    print()

game_log
Size is (171907, 161)
Columns: Index(['date', 'number_of_game', 'day_of_week', 'v_name', 'v_league',
       'v_game_number', 'h_name', 'h_league', 'h_game_number', 'v_score',
       ...
       'h_player_7_name', 'h_player_7_def_pos', 'h_player_8_id',
       'h_player_8_name', 'h_player_8_def_pos', 'h_player_9_id',
       'h_player_9_name', 'h_player_9_def_pos', 'additional_info',
       'acquisition_info'],
      dtype='object', length=161)

park
Size is (252, 9)
Columns: Index(['park_id', 'name', 'aka', 'city', 'state', 'start', 'end', 'league',
       'notes'],
      dtype='object')

person
Size is (20494, 7)
Columns: Index(['id', 'last', 'first', 'player_debut', 'mgr_debut', 'coach_debut',
       'ump_debut'],
      dtype='object')

team
Size is (150, 8)
Columns: Index(['team_id', 'league', 'start', 'end', 'city', 'nickname', 'franch_id',
       'seq'],
      dtype='object')



### Game Log

In [6]:
game_log.shape

(171907, 161)

In [7]:
game_log.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 171907 entries, 0 to 171906
Columns: 161 entries, date to acquisition_info
dtypes: float64(77), int64(6), object(78)
memory usage: 211.2+ MB


In [8]:
game_log.describe()

date  number_of_game  v_game_number  h_game_number  \
count  1.719070e+05   171907.000000  171907.000000  171907.000000   
mean   1.953462e+07        0.260897      76.929887      76.953806   
std    4.149326e+05        0.605667      45.178029      45.162564   
min    1.871050e+07        0.000000       1.000000       1.000000   
25%    1.918052e+07        0.000000      38.000000      38.000000   
50%    1.953053e+07        0.000000      76.000000      76.000000   
75%    1.989051e+07        0.000000     115.000000     115.000000   
max    2.016100e+07        3.000000     165.000000     165.000000   

             v_score        h_score    length_outs     attendance  \
count  171907.000000  171907.000000  140841.000000  118877.000000   
mean        4.420582       4.701461      53.619976   20184.247188   
std         3.278489       3.355605       5.571512   14257.381902   
min         0.000000       0.000000       0.000000       0.000000   
25%         2.000000       2.000000      51.000000    7962.000000   
50%         4.000000       4.000000      54.000000   18639.000000   
75%         6.000000       6.000000      54.000000   31242.000000   
max        49.000000      38.000000     156.000000   99027.000000   

       length_minutes      v_at_bats         v_hits      v_doubles  \
count   136701.000000  140838.000000  140838.000000  140772.000000   
mean       150.903329      34.914398       8.999318       1.563592   
std         34.748160       4.633761       3.599728       1.346740   
min         24.000000      12.000000       0.000000       0.000000   
25%        125.000000      32.000000       6.000000       1.000000   
50%        150.000000      34.000000       9.000000       1.000000   
75%        173.000000      37.000000      11.000000       2.000000   
max       1245.000000      85.000000      42.000000      12.000000   

           v_triples     v_homeruns          v_rbi  v_sacrifice_hits  \
count  140835.000000  140800.000000  139488.000000     140838.000000   
mean        0.276039       0.729119       4.041366          0.558741   
std         0.560388       0.959742       3.021488          0.862961   
min        -1.000000      -1.000000       0.000000          0.000000   
25%         0.000000       0.000000       2.000000          0.000000   
50%         0.000000       0.000000       3.000000          0.000000   
75%         0.000000       1.000000       6.000000          1.000000   
max         6.000000       8.000000      31.000000          9.000000   

       v_sacrifice_flies  v_hit_by_pitch        v_walks  v_intentional_walks  \
count      135885.000000   140838.000000  140728.000000        113541.000000   
mean            0.169746        0.236428       3.153409             0.246757   
std             0.437917        0.508436       2.128045             0.544284   
min             0.000000        0.000000       0.000000            -1.000000   
25%             0.000000        0.000000       2.000000             0.000000   
50%             0.000000        0.000000       3.000000             0.000000   
75%             0.000000        0.000000       4.000000             0.000000   
max             5.000000        6.000000      18.000000             7.000000   

        v_strikeouts  v_stolen_bases  v_caught_stealing  \
count  140746.000000   140775.000000      127317.000000   
mean        5.290381        0.583342           0.299002   
std         3.012270        0.934292           0.562095   
min        -1.000000       -1.000000          -1.000000   
25%         3.000000        0.000000           0.000000   
50%         5.000000        0.000000           0.000000   
75%         7.000000        1.000000           1.000000   
max        26.000000       11.000000           6.000000   

       v_grounded_into_double  v_first_catcher_interference  v_left_on_base  \
count           131222.000000                 134357.000000   140838.000000   
mean                 0.602529                      0.003520        7.106697   
s

In [9]:
game_log.head()

date  number_of_game day_of_week v_name v_league  v_game_number h_name  \
0  18710504               0         Thu    CL1      NaN              1    FW1   
1  18710505               0         Fri    BS1      NaN              1    WS3   
2  18710506               0         Sat    CL1      NaN              2    RC1   
3  18710508               0         Mon    CL1      NaN              3    CH1   
4  18710509               0         Tue    BS1      NaN              2    TRO   

  h_league  h_game_number  v_score  h_score  length_outs day_night completion  \
0      NaN              1        0        2         54.0         D        NaN   
1      NaN              1       20       18         54.0         D        NaN   
2      NaN              1       12        4         54.0         D        NaN   
3      NaN              1       12       14         54.0         D        NaN   
4      NaN              1        9        5         54.0         D        NaN   

  forefeit protest park_id  attendance  length_minutes v_line_score  \
0      NaN     NaN   FOR01       200.0           120.0    000000000   
1      NaN     NaN   WAS01      5000.0           145.0    107000435   
2      NaN     NaN   RCK01      1000.0           140.0    610020003   
3      NaN     NaN   CHI01      5000.0           150.0    101403111   
4      NaN     NaN   TRO01      3250.0           145.0    000002232   

  h_line_score  v_at_bats  v_hits  v_doubles  v_triples  v_homeruns  v_rbi  \
0    010010000       30.0     4.0        1.0        0.0         0.0    0.0   
1    640113030       41.0    13.0        1.0        2.0         0.0   13.0   
2    010020100       49.0    11.0        1.0        1.0         0.0    8.0   
3    077000000       46.0    15.0        2.0        1.0         2.0   10.0   
4    101003000       46.0    17.0        4.0        1.0         0.0    6.0   

   v_sacrifice_hits  v_sacrifice_flies  v_hit_by_pitch  v_walks  \
0               0.0                0.0             0.0      1.0   
1               0.0                0.0             0.0     18.0   
2               0.0                0.0             0.0      0.0   
3               0.0                0.0             0.0      0.0   
4               0.0                0.0             0.0      2.0   

   v_intentional_walks  v_strikeouts  v_stolen_bases  v_caught_stealing  \
0                  NaN           6.0             1.0                NaN   
1                  NaN           5.0             3.0                NaN   
2                  NaN           1.0             0.0                NaN   
3                  NaN           1.0             0.0                NaN   
4                  NaN           0.0             1.0                NaN   

   v_grounded_into_double  v_first_catcher_interference  v_left_on_base  \
0                    -1.0                           NaN             4.0   
1                    -1.0                           NaN            12.0   
2                    -1.0                           NaN            10.0   
3                    -1.0                           NaN             7.0   
4                    -1.0                           NaN            12.0   

   v_pitchers_used  v_individual_earned_runs  v_team_earned_runs  \
0              1.0                       1.0                 1.0   
1              1.0                       6.0                 6.0   
2              1.0                       0.0                 0.0   
3              1.0                       6.0                 6.0   
4              1.0                       2.0                 2.0   

   v_wild_pitches  v_balks  v_putouts  v_assists  v_errors  v_passed_balls  \
0             0.0      0.0       27.0        9.0       0.0             3.0   
1             1.0      0.0       27.0       13.0      10.0             1.0   
2             2.0      0.0       27.0       12.0       8.0             5.0   
3             0.0      0.0       27.0       15.0      11.0             6.0   
4             0.0      0.0       

In [10]:
game_log.tail()

date  number_of_game day_of_week v_name v_league  v_game_number  \
171902  20161002               0         Sun    MIL       NL            162   
171903  20161002               0         Sun    NYN       NL            162   
171904  20161002               0         Sun    LAN       NL            162   
171905  20161002               0         Sun    PIT       NL            162   
171906  20161002               0         Sun    MIA       NL            161   

       h_name h_league  h_game_number  v_score  h_score  length_outs  \
171902    COL       NL            162        6        4         60.0   
171903    PHI       NL            162        2        5         51.0   
171904    SFN       NL            162        1        7         51.0   
171905    SLN       NL            162        4       10         51.0   
171906    WAS       NL            162        7       10         51.0   

       day_night completion forefeit protest park_id  attendance  \
171902         D        NaN      NaN     NaN   DEN02     27762.0   
171903         D        NaN      NaN     NaN   PHI13     36935.0   
171904         D        NaN      NaN     NaN   SFO03     41445.0   
171905         D        NaN      NaN     NaN   STL10     44615.0   
171906         D        NaN      NaN     NaN   WAS11     28730.0   

        length_minutes v_line_score h_line_score  v_at_bats  v_hits  \
171902           203.0   0200000202   1100100010       39.0    10.0   
171903           159.0    000001100    00100031x       33.0     8.0   
171904           184.0    000100000    23000002x       30.0     4.0   
171905           192.0    000020200    00100360x       35.0     9.0   
171906           216.0    000230020    03023002x       38.0    14.0   

        v_doubles  v_triples  v_homeruns  v_rbi  v_sacrifice_hits  \
171902        4.0        1.0         2.0    6.0               0.0   
171903        3.0        0.0         0.0    2.0               0.0   
171904        0.0        0.0         0.0    1.0               0.0   
171905        0.0        0.0         1.0    4.0               0.0   
171906        1.0        1.0         2.0    7.0               1.0   

        v_sacrifice_flies  v_hit_by_pitch  v_walks  v_intentional_walks  \
171902                0.0             1.0      4.0                  0.0   
171903                0.0             0.0      2.0                  0.0   
171904                0.0             0.0      2.0                  0.0   
171905                0.0             0.0      4.0                  0.0   
171906                0.0             0.0      3.0                  2.0   

        v_strikeouts  v_stolen_bases  v_caught_stealing  \
171902          12.0             2.0                1.0   
171903           9.0             1.0                1.0   
171904           7.0             0.0                0.0   
171905          11.0             0.0                1.0   
171906          10.0             1.0                1.0   

        v_grounded_into_double  v_first_catcher_interference  v_left_on_base  \
171902                     0.0                           0.0             8.0   
171903                     1.0                           0.0             6.0   
171904                     1.0                           0.0             4.0   
171905                     0.0                           0.0             8.0   
171906                     1.0                           0.0             8.0   

        v_pitchers_used  v_individual_earned_runs  v_team_earned_runs  \
171902              7.0                       4.0                 4.0   
171903              6.0                       3.0                 3.0   
171904              7.0                       7.0                 7.0   
171905              6.0                       8.0                 8.0   
171906              7.0                      10.0                10.0   

        v_wild_pitches  v_balks  v_putouts  v_assists  v_errors  \
171902             1.0      0.0       30.0       12.0       

In [11]:
game_log.columns

Index(['date', 'number_of_game', 'day_of_week', 'v_name', 'v_league',
       'v_game_number', 'h_name', 'h_league', 'h_game_number', 'v_score',
       ...
       'h_player_7_name', 'h_player_7_def_pos', 'h_player_8_id',
       'h_player_8_name', 'h_player_8_def_pos', 'h_player_9_id',
       'h_player_9_name', 'h_player_9_def_pos', 'additional_info',
       'acquisition_info'],
      dtype='object', length=161)

In [12]:
!ls

appearance_type.csv
game_log.csv
game_log_fields.txt
park_codes.csv
person_codes.csv
project13_designing_and_creating_a_database.ipynb
team_codes.csv


In [13]:
!cat game_log_fields.txt

Field(s)  Meaning
    1     Date in the form "yyyymmdd"
    2     Number of game:
             "0" -- a single game
             "1" -- the first game of a double (or triple) header
                    including seperate admission doubleheaders
             "2" -- the second game of a double (or triple) header
                    including seperate admission doubleheaders
             "3" -- the third game of a triple-header
             "A" -- the first game of a double-header involving 3 teams
             "B" -- the second game of a double-header involving 3 teams
    3     Day of week  ("Sun","Mon","Tue","Wed","Thu","Fri","Sat")
  4-5     Visiting team and league
    6     Visiting team game number
          For this and the home team game number, ties are counted as
          games and suspended games are counted from the starting
          rather than the ending date.
  7-8     Home team and league
    9     Home team game number
10-11     Visiting and home team score (unquoted)


### Park Codes

In [14]:
park.shape

(252, 9)

In [15]:
park.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 252 entries, 0 to 251
Data columns (total 9 columns):
park_id    252 non-null object
name       252 non-null object
aka        58 non-null object
city       252 non-null object
state      252 non-null object
start      252 non-null object
end        222 non-null object
league     186 non-null object
notes      128 non-null object
dtypes: object(9)
memory usage: 17.8+ KB


In [16]:
park.describe()

park_id           name                  aka          city state  \
count      252            252                   58           252   252   
unique     252            241                   56            85    36   
top      NYC03  Athletic Park  Federal League Park  Philadelphia    NY   
freq         1              4                    2            14    40   

             start         end league notes  
count          252         222    186   128  
unique         215         205      6   123  
top     05/01/1883  10/03/1915     NL   PIT  
freq             5           3     88     2

In [17]:
park.head()

park_id                           name  \
0   ALB01                 Riverside Park   
1   ALT01                  Columbia Park   
2   ANA01       Angel Stadium of Anaheim   
3   ARL01              Arlington Stadium   
4   ARL02  Rangers Ballpark in Arlington   

                                        aka       city state       start  \
0                                       NaN     Albany    NY  09/11/1880   
1                                       NaN    Altoona    PA  04/30/1884   
2             Edison Field; Anaheim Stadium    Anaheim    CA  04/19/1966   
3                                       NaN  Arlington    TX  04/21/1972   
4  The Ballpark in Arlington; Ameriquest Fl  Arlington    TX  04/11/1994   

          end league                                           notes  
0  05/30/1882     NL  TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1  05/31/1884     UA                                             NaN  
2         NaN     AL                                             NaN  
3  10/03/1993     AL                                             NaN  
4         NaN     AL                                             NaN

In [18]:
park.tail()

park_id                                 name  aka           city state  \
247   WIL02        BB&T Ballpark at Bowman Field  NaN    Wiliamsport    PA   
248   WNY01     West New York Field Club Grounds  NaN  West New York    NJ   
249   WOR01   Agricultural County Fair Grounds I  NaN      Worcester    MA   
250   WOR02  Agricultural County Fair Grounds II  NaN      Worcester    MA   
251   WOR03       Worcester Driving Park Grounds  NaN      Worcester    MA   

          start         end league  \
247  08/20/2017  08/20/2017     NL   
248  09/11/1898  09/17/1899     NL   
249  05/01/1880  09/29/1882     NL   
250  08/17/1887  08/17/1887     NL   
251  10/30/1874  10/30/1874    NaN   

                                                      notes  
247                                                     PIT  
248  BRO:9/18&10/2/1898; NY1:9/11/98, 6/4&7/16&8/13&9/17/99  
249                                                     NaN  
250                                              1 BSN game  
251                                              1 BS1 game

In [19]:
park.columns

Index(['park_id', 'name', 'aka', 'city', 'state', 'start', 'end', 'league',
       'notes'],
      dtype='object')

### Person Codes

In [20]:
person.shape

(20494, 7)

In [21]:
person.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20494 entries, 0 to 20493
Data columns (total 7 columns):
id              20494 non-null object
last            20494 non-null object
first           20433 non-null object
player_debut    19183 non-null object
mgr_debut       701 non-null object
coach_debut     1566 non-null object
ump_debut       1546 non-null object
dtypes: object(7)
memory usage: 1.1+ MB


In [22]:
person.describe()

id   last  first player_debut   mgr_debut coach_debut  \
count      20494  20494  20433        19183         701        1566   
unique     20494  10397   2484        10278         537         332   
top     martm101  Smith   Bill   05/01/1884  05/01/1884  04/01/2013   
freq           1    168    579           36          10          29   

         ump_debut  
count         1546  
unique        1130  
top     08/25/1978  
freq            52

In [23]:
person.head()

id     last   first player_debut mgr_debut coach_debut ump_debut
0  aardd001  Aardsma   David   04/06/2004       NaN         NaN       NaN
1  aaroh101    Aaron    Hank   04/13/1954       NaN         NaN       NaN
2  aarot101    Aaron  Tommie   04/10/1962       NaN  04/06/1979       NaN
3  aased001     Aase     Don   07/26/1977       NaN         NaN       NaN
4  abada001     Abad    Andy   09/10/2001       NaN         NaN       NaN

In [24]:
person.tail()

id      last   first player_debut mgr_debut coach_debut ump_debut
20489  zuvep001   Zuvella    Paul   09/04/1982       NaN  04/02/1996       NaN
20490  zuveg101  Zuverink  George   04/21/1951       NaN         NaN       NaN
20491  zwild101  Zwilling   Dutch   08/14/1910       NaN  04/15/1941       NaN
20492  zycht001      Zych    Tony   09/04/2015       NaN         NaN       NaN
20493  thoma102  Thompson     NaN          NaN       NaN         NaN       NaN

In [25]:
person.columns

Index(['id', 'last', 'first', 'player_debut', 'mgr_debut', 'coach_debut',
       'ump_debut'],
      dtype='object')

### Team Codes

In [26]:
team.shape

(150, 8)

In [27]:
team.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 8 columns):
team_id      150 non-null object
league       124 non-null object
start        150 non-null int64
end          150 non-null int64
city         150 non-null object
nickname     150 non-null object
franch_id    150 non-null object
seq          150 non-null int64
dtypes: int64(3), object(5)
memory usage: 9.5+ KB


In [28]:
team.describe()

start          end         seq
count   150.000000   150.000000  150.000000
mean   1902.733333  1517.440000    1.293333
std      37.326002   761.772866    0.585584
min    1871.000000     0.000000    1.000000
25%    1879.000000  1872.000000    1.000000
50%    1887.000000  1884.000000    1.000000
75%    1911.250000  1891.000000    1.000000
max    2012.000000  2011.000000    4.000000

In [29]:
team.head()

team_id league  start   end       city         nickname franch_id  seq
0     ALT     UA   1884  1884    Altoona  Mountain Cities       ALT    1
1     ARI     NL   1998     0    Arizona     Diamondbacks       ARI    1
2     BFN     NL   1879  1885    Buffalo           Bisons       BFN    1
3     BFP     PL   1890  1890    Buffalo           Bisons       BFP    1
4     BL1    NaN   1872  1874  Baltimore         Canaries       BL1    1

In [30]:
team.tail()

team_id league  start   end        city   nickname franch_id  seq
145     WS8     NL   1886  1889  Washington   Senators       WS8    1
146     WS9     AA   1891  1891  Washington   Senators       WS9    1
147     WSN     NL   1892  1899  Washington   Senators       WS9    2
148     WSU     UA   1884  1884  Washington  Nationals       WSU    1
149     MIA     NL   2012     0       Miami    Marlins       FLO    2

In [31]:
team.columns

Index(['team_id', 'league', 'start', 'end', 'city', 'nickname', 'franch_id',
       'seq'],
      dtype='object')

**Observations**

Game Log contains all the information about baseball games played from the 1800s through to today. This include play statistics for both teams, conditions of the game including location, attendance...etc.

Park Codes contains information about the Baseball stadiums - Park ID, name, location, league, when they started to be used and when they were no longer used. The park_id seems to correspond to the park_id in Game Logs.

Person Codes contains information about the individual players, managers, coaches and umpires. It includes ID, First and Last name along with a debut date - for player, manager, coach and empire respectively. Some people were both players and later managers and coaches. The ID corresponds to a several columns in Game Logs (1b_umpire_id, manager_id, winning_pitch_id...etc).

Team Codes contains information about the respective teams. Their ID, which league they belonged to, when the team started and ended, City they were based in, their nickname, franchise ID along with seq. Not quite sure what the latest column (seq) is for. The team_id seems to correspond to the v_name/h_name in Game Logs.

**Research**

_What each defensive position number represents?_

1 - pitcher
2 - catcher
3 - first baseman
4 - second baseman
5 - third baseman
6 - shortstop
7 - left fielder
8 - center fielder
9 - right fielder
10 - not sure?

_The values in the various league fields and which leagues they represent_

NL - National League
AL - American League
AA - American Association
UA - Union Association (1884)
FL - Federal League (1914–1915)
PL - Player's League (1890)

In [32]:
team.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 8 columns):
team_id      150 non-null object
league       124 non-null object
start        150 non-null int64
end          150 non-null int64
city         150 non-null object
nickname     150 non-null object
franch_id    150 non-null object
seq          150 non-null int64
dtypes: int64(3), object(5)
memory usage: 9.5+ KB


In [33]:
# unique leagues
team['league'].value_counts(dropna=False)

NL     45
NaN    26
AL     25
AA     24
UA     13
FL      9
PL      8
Name: league, dtype: int64

In [34]:
# unique defensive positions
game_log['h_player_6_def_pos'].value_counts(dropna=False)

NaN      31069
 5.0     24202
 3.0     21125
 2.0     18973
 9.0     18766
 7.0     17520
 8.0     13198
 4.0     12401
 6.0     10192
 10.0     4401
 1.0        60
Name: h_player_6_def_pos, dtype: int64

In [35]:
# unique defensive positions
game_log['h_player_7_def_pos'].value_counts(dropna=False)

NaN      31069
 2.0     30064
 6.0     23739
 5.0     23049
 4.0     19043
 3.0     12200
 9.0     10342
 8.0     10193
 7.0      9610
 10.0     2550
 1.0        48
Name: h_player_7_def_pos, dtype: int64

---

## Importing Data into SQLite

---

## Looking for Normalization Opportunities

---

## Planning a Normalized Schema

---

## Creating Tables Without Foreign Key Relations

---

## Adding the Team and Game Tables

---

## Adding the Team Appearance Table

---

## Adding the Person Appearance Table

---

## Removing the Original Tables

---

## Next Steps